### ISSUE TITLE
parse with format "yyMd" gives unexpected result

### ISSUE THREAD

**Issue by nicklasisraelsson** [date-fns/date-fns#2973]
I'm unsure if my expectations are off but when I parse the string "2211" and "2212" using format "yyMd" I don't get a valid date. "2213" works.

```js
  const actualDate = parse("2211", "yyMd", new Date());
  console.log(actualDate.toString());
  //Invalid Date

  const expectedDate = new Date("2022-01-01");
  console.log(expectedDate.toString());
  //Sat Jan 01 2022 01:00:00 GMT+0100 (Central European Standard Time)
```

**Comment by HenriqueDerosa** [comment]
Hey @nicklasisraelsson,

It seems the Month in those cases cannot be a single character for two reasons:
1. there's no separation between month character and day character, which impacts the following 👇🏻 
2. when you use `2211` the parse is assuming you specified year `22`(2 characters) and month  `11`(2 characters). So it's missing the `d`. Same applies to `2212`. 

It works with `2213` because there's no month 13.

I think this date format `yyMd` is prone to this kind of problem.

**Comment by nicklasisraelsson** [comment]
Yeah I agree that the `yyMd` format is probably prone to these errors. 

I was however surprised that `parse` rather choose to return an invalid date since it can match `11` to a month with no day and didn't try matching the `11` to January 1st.

**Comment by Alanscut** [comment]
![image](https://user-images.githubusercontent.com/21104054/160523132-1cfafe78-3e38-4697-8247-67cda9616bef.png)
I think the format like `yyMd ` is too complicated to parse accurately.
Such as `yyMdHms`-->`221111111`, who knows what time this is?  It can be 2022-1-11 1:11:1 or 2022-11-1 11:1:1, etc…. 
There are too many possibilities, so it's better to use '/'(or other character) as separation, `yy/M/d/H/m/s`-->`22/1/11/1/11/1`

**Comment by nicklasisraelsson** [comment]
I agree that format `yyMdHms` is an even _more_ complicated format. But in that case I would expect it would take the first valid date using the priority of the different units that matched and return that.

But as I mentioned before, I was not expecting `parse` to return `Invalid Date` for an input that what I thought was valid. I was expecting `parse` to re-try `11` as `1st of January` when parsing `11` as  `0th of November` resulted in an invalid date.

But it might very well be that my expectations are wrong.

**Comment by Alanscut** [comment]
```
  const date_temp = new Date('2022-1-11')
  console.log(date_temp)
  const format_temp = format(date_temp,'yyMd')
  console.log(format_temp)
  console.log(parse(format_temp,'yyMd',new Date()))
```
get
```
Tue Jan 11 2022 00:00:00 GMT+0000
22111
Tue Nov 01 2022 00:00:00 GMT+0000 
//the date use both format() and parse() finally get a different result
```
my point is we should avoid use format like `yyMd`
maybe here need a warn to users ? what do you think?

**Comment by nicklasisraelsson** [comment]
I see that you've used the `11th of January` in your example instead of the `1st of January`. If we use `format` with `1st of January` we get `2211`. And we know what happens when we try to parse that. 😃 

I think `parse` is doing as expected when parsing format `yyMd` with an string like `22111` into `1st of November` since the `M` unit has higher priority than `d`. This is probably also why we get `Invalid Date` for `2211`.

I don't know the code and wrongfully assumed that it had some way of handling ambiguous strings where one interpretation resulted in `Invalid Date`. That there was something that after parsing the `11` as the `M` it would retry with interpreting the `11` as `Md` since `1` is a valid value for `M` too. 

I opened this issue to know if this is considered a bug or just an accepted side effect of using a "fuzzy" format like `yyMd`.

**Comment by Alanscut** [comment]
my point is this is just an accepted side effect of using a "fuzzy" format like `yyMd`, 
this is just my personal opinion，maybe we should @ the member of this lib to confirm the situation
